### PR TITLE
multi dust

### LINKED ISSUE #92: Multiparameter "Dust" object
In order to support both SMC2 and multiple chains at once. Currently when we create a dust object we do:

```r
dust::dust$new(pars, step, n_particles)
```

but this would become 

```r
dust::dust_multi(pars, index, step, n_particles)
```

where `pars` is an unnamed list of named parameters, and `index` is a mapping indicating how they should be laid out. This might look like `rep(1:3, each = 100)` for 3 chains with 100 particles each, for example.

The construction is fairly straightforward; see the `dust_simulate` code for an implementation.

We need to support all the other methods in a consistent way though:

* `run`: does this return the same size or do we return an extra rank on the array? This depends if we always need the same number of particles per replicate. That is at least true if we want to use this to run multiple chains of the pmcmc, but is likely to be too limiting.
* `set_state`, `state`: reflect return value of `run`
* `reset`: require the same size data as for initialisation, unless the indexes change?
* `reorder`: if we just treat the list as one big list this is trivial
* `info`: returns a list
* `data`: returns a list

Unaffected: `name`, `set_index`, `index`, `n_threads`, `rng_state`

We also need tor return a the parameter mapping, and clearly need a better name than `index` as that is used for the state index.

An alternative approach is that we require the user to do the expansion for us, as with `dust_simulate`, but that makes the info/data methods near useless.

### PR DESCRIPTION
**Pull request by richfitz**
Start on support for multi-dust objects.

From the point of view code at a lower level than interface.hpp there's almost no change: there is a longer vector of particles and a couple of overloaded methods, but otherwise it's all hidden in the private interface. The interface.hpp file does most of the heavy lifting for the changes.

Fixes #92

### CODE REVIEW

**Review by johnlees — CHANGES_REQUESTED** [review]
Nice changes to the interface which should make the CUDA implementation as easy as possible 🙂

File: DESCRIPTION [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Package: dust
 Title: Iterate Multiple Realisations of Stochastic Models
-Version: 0.5.8
+Version: 0.5.9
```
[the remark this answers is not in the document]
Comment (richfitz) [reply DESCRIPTION:3]: yeah - can do that now perhaps as most of the really tricky parts are in mcstate

File: NEWS.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
+# dust 0.5.9
+
+* Start of support for running dust objects with multiple data/parameter sets at once (#92)
+
```
[the remark this answers is not in the document]
Comment (richfitz) [reply NEWS.md:4]: Was not user-visible (the shared memory example)

File: inst/include/dust/interface.hpp [hunk 3]
```diff
@@ -170,37 +191,101 @@ void dust_set_state(Dust<T> *obj, cpp11::doubles_matrix r_state) {
...
+  }
+
+  const std::vector<typename T::real_t> state(r_state.begin(), r_state.end());
+
+  obj->set_state(state, true);
+}
+
 template <typename T>
-cpp11::writable::doubles_matrix dust_run(SEXP ptr, int step_end) {
+cpp11::sexp dust_run(SEXP ptr, int step_end) {
```
[the remark this answers is not in the document]
Comment (richfitz) [reply inst/include/dust/interface.hpp:231]: I had compilation errors on clang due to this, not sure why. No disadvantage as they're all SEXPs ultimately (it's not a coercion as such, so it's more just naming the base type)

File: inst/include/dust/interface.hpp [hunk 4]
```diff
@@ -23,29 +23,46 @@ inline void validate_positive(int x, const char *name);
...
                        cpp11::sexp r_seed) {
   validate_size(step, "step");
   validate_positive(n_particles, "n_particles");
   validate_positive(n_threads, "n_threads");
   std::vector<uint64_t> seed = as_rng_seed<typename T::real_t>(r_seed);
 
-  typename T::init_t data = dust_data<T>(r_data);
-
-  Dust<T> *d = new Dust<T>(data, step, n_particles, n_threads, seed);
+  Dust<T> *d = NULL;
```
[the remark this answers is not in the document]
Comment (richfitz) [reply inst/include/dust/interface.hpp:49]: explicit nullptr is better IMO. I always forget that `NULL` is not the proper C++ way to do this

File: inst/include/dust/dust.hpp [hunk 5]
```diff
@@ -266,19 +292,53 @@ class Dust {
   }
 
 private:
-  std::vector<size_t> _index;
+  const size_t _n_data; // 0 in the "single" case, >=1 otherwise
+  const size_t _n_particles_total; // Total number of particles
   size_t _n_threads;
   dust::pRNG<real_t> _rng;
+
+  std::vector<size_t> _index;
```
[the remark this answers is not in the document]
Comment (richfitz) [reply inst/include/dust/dust.hpp:300]: Can be inferred as it's always balanced. Why will you need this in CUDA?

File: inst/include/dust/interface.hpp [hunk 6]
```diff
@@ -170,37 +191,101 @@ void dust_set_state(Dust<T> *obj, cpp11::doubles_matrix r_state) {
...
+    typename T::init_t data = dust_data<T>(r_data);
+    obj->set_data(data);
+    return dust_info<T>(data);
+  } else {
+    // The underlying implementation should be tidied up, as the
+    // single case leaves us with inconsistent data already, and the
+    // error management is tricky (#125)
+    cpp11::stop("set_data() with data_multi not yet supported");
+  }
+  return info;
```
[the remark this answers is not in the document]
Comment (richfitz) [reply inst/include/dust/interface.hpp:288]: oops, removed the early exit now

File: inst/include/dust/interface.hpp [hunk 7]
```diff
@@ -346,6 +456,37 @@ inline std::vector<size_t> r_index_to_index(cpp11::sexp r_index, size_t nmax) {
   return index;
 }
 
+inline std::vector<size_t> r_index_reorder_matrix(cpp11::sexp r_index,
+                                                  const size_t n_particles,
+                                                  const size_t n_data) {
```
[the remark this answers is not in the document]
Comment (richfitz) [reply inst/include/dust/interface.hpp:461]: yes, but it's subject to the constraint that the number of particles stays the same - I've added an easier to understand test in f475726

File: inst/include/dust/dust.hpp [hunk 8]
```diff
@@ -95,18 +95,40 @@ class Dust {
   typedef typename T::init_t init_t;
   typedef typename T::real_t real_t;
 
-  Dust(const init_t data, const size_t step, const size_t n_particles,
+  // Can we do this with
+  // template <typename U>
+  // Dust(const U& data, ...)
+  // and having all the overloading in initialise? that would seem ideal
+  // really.
```
[the remark this answers is not in the document]
Comment (richfitz) [reply inst/include/dust/dust.hpp:102]: ok cool, removed the comment

File: inst/include/dust/dust.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (johnlees) [reply inst/include/dust/dust.hpp:295]: shared_t will be loaded into __shared__, but the shared_t in question will depend on which data the particle belongs to. So we just need a relation between thread/block/grid and data index

[the remark this answers is not in the document]
Comment (johnlees) [reply inst/include/dust/dust.hpp:295]: We don't currently use the concept of blocks particularly, but they become more useful when you have memory shared between groups of threads.
Example getting a thread and block index: https://github.com/johnlees/pp-sketchlib/blob/master/src/gpu/dist.cu#L171-L197
Using it to load shared memory: https://github.com/johnlees/pp-sketchlib/blob/master/src/gpu/dist.cu#L217-L231

**Review by johnlees — APPROVED** [review]

**Pull request merged**